modules/volume/volume.go

```diff
@@ -190,13 +147,118 @@ func DefaultClickHandler(v Volume, c Controller, e bar.Event) {
 // Stream starts the module.
 func (m *Module) Stream() <-chan bar.Output {
 	ch := base.NewChannel()
-	go m.worker()
+	go m.runWorker()
 	go m.outputLoop(ch)
 	return ch
 }
 
+// init initializes the Module struct to usable default values.
+func (m *Module) init() {
+	l.Register(m, "outputFunc", "currentVolume", "clickHandler")
+	m.OnClick(DefaultClickHandler)
+	// Default output template is just the volume %, "MUT" when muted.
+	m.OutputTemplate(outputs.TextTemplate(`{{if .Mute}}MUT{{else}}{{.Pct}}%{{end}}`))
+}
+
+// runWorker runs the worker from the moduleImpl interface, and sets
+// the error on currentVolume if an error results.
+func (m *Module) runWorker() {
+	m.currentVolume.Error(m.impl.worker())
+}
+
+// outputLoop listens for updates to the volume, as well as the output
+// function,
+// and updates the module output.
+func (m *Module) outputLoop(ch base.Channel) {
+	v, err := m.currentVolume.Get()
+	sVol := m.currentVolume.Subscribe()
+
+	outputFunc := m.outputFunc.Get().(func(Volume) bar.Output)
+	sOutputFunc := m.outputFunc.Subscribe()
+
+	for {
+		if ch.Error(err) {
+			return
+		}
+		if vol, ok := v.(Volume); ok {
+			ch.Output(outputFunc(vol))
+		}
+		select {
+		case <-sVol:
+			v, err = m.currentVolume.Get()
+		case <-sOutputFunc:
+			outputFunc = m.outputFunc.Get().(func(Volume) bar.Output)
+		}
+	}
+}
+
+// ALSA implementation.
+
+type alsaModule struct {
+	Module
+	cardName  string
+	mixerName string
+
+	// To make it easier to change volume using alsa apis,
+	// store the current state and snd_mixer_elem_t pointer.
+	elem *C.snd_mixer_elem_t
+}
+
+// Mixer constructs an instance of the volume module for a
+// specific card and mixer on that card.
+func Mixer(card, mixer string) *Module {
+	m := &alsaModule{
+		cardName:  card,
+		mixerName: mixer,
+	}
+	m.impl = m
```

Comment:
This is a cyclic reference: alsaModule has a Module, but that Module points to alsaModule as the impl.
Is there a way we can avoid this loop?

What if instead of worker() error, the impl returned a chan struct{Volume, error}?
It seems like the main reason for the cycle is that impl needs access to Module's "currentVolume", so using a channel would eliminate that.

Some of the SetVolume code could also be moved into module, and removed from impl. Instead impl could have SetVolume(newVol int64) error, and if error is non-nil, Module assumes update was successful and updates its stored volume.

Reply:
What's wrong with a cyclic reference? I'm effectively using the interface as a manual vtable, to let the base implementation call out to subclass-specific implementations.

Note that making worker() return a chan won't work either - the OnClick handler (in the base module) needs to get access to a Controller, which has implementation-specific parts.

Reply:
A Module has an implementation, so Module { Impl } is fine. It's the impl->Module that concerns me, mostly because it makes the code slightly harder to follow. Both pa/alsa need to remember to call init() on their m, and instead of returning a paModule, they're returning the embedded module, which at first glance makes it seem like the paModule is being discarded even though it's not.

Impl just needs to:
- emit the current volume,
- and set the volume/muted.
Module can continue to implement Controller, with SetVolume/Muted calling impl.setVolume/Muted -> error, and updating currentVolume for nil errors.

If the channel struct{Volume, error} is verbose, we could use `worker(update func(Volume)) error` in impl, and have Module.Stream call it with `impl.worker(func(v Volume) {m.currentVolume.Set(v)})`.